File: src/main/java/io/github/homchom/recode/mod/features/TemplatePeeker.java
/*
package io.github.homchom.recode.mod.features;

import com.google.gson.*;
import com.mojang.blaze3d.vertex.PoseStack;
import io.github.homchom.recode.Recode;
import io.github.homchom.recode.mod.config.Config;
import io.github.homchom.recode.sys.file.ILoader;
import io.github.homchom.recode.sys.hypercube.codeaction.*;
import io.github.homchom.recode.sys.hypercube.templates.*;
import io.github.homchom.recode.sys.renderer.block.BlockRenderer;
import net.fabricmc.fabric.api.client.rendering.v1.WorldRenderEvents;
import net.fabricmc.loader.api.FabricLoader;
import net.minecraft.client.Minecraft;
import net.minecraft.core.*;
import net.minecraft.network.chat.TextComponent;
import net.minecraft.world.level.block.*;
import net.minecraft.world.level.block.entity.*;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.properties.BlockStateProperties;
import net.minecraft.world.phys.Vec3;

import java.awt.*;
import java.lang.reflect.Field;
import java.util.*;

public class TemplatePeeker implements ILoader {

    public static boolean templatePreview = false;
    public static boolean templateFits = true;
    public static HashMap<String, Block> blockTypes = new HashMap<>();
    public static HashSet<String> noStones = new HashSet<>();
    public static HashSet<String> noChests = new HashSet<>();
    private static BlockRenderer blockRenderer = new BlockRenderer();

    @Override
    public void load() {

        blockTypes.put("func", Blocks.LAPIS_BLOCK);
        blockTypes.put("set_var", Blocks.IRON_BLOCK);
        blockTypes.put("event", Blocks.DIAMOND_BLOCK);
        blockTypes.put("player_action", Blocks.COBBLESTONE);
        blockTypes.put("if_player", Blocks.OAK_PLANKS);
        blockTypes.put("else", Blocks.END_STONE);
        blockTypes.put("call_func", Blocks.LAPIS_ORE);
        blockTypes.put("entity_event", Blocks.GOLD_BLOCK);
        blockTypes.put("process", Blocks.EMERALD_BLOCK);
        blockTypes.put("repeat", Blocks.PRISMARINE);
        blockTypes.put("entity_action", Blocks.MOSSY_COBBLESTONE);
        blockTypes.put("start_process", Blocks.EMERALD_ORE);
        blockTypes.put("if_var", Blocks.OBSIDIAN);
        blockTypes.put("if_entity", Blocks.BRICKS);
        blockTypes.put("control", Blocks.COAL_BLOCK);
        blockTypes.put("select_obj", Blocks.PURPUR_BLOCK);
        blockTypes.put("if_game", Blocks.RED_NETHER_BRICKS);
        blockTypes.put("game_action", Blocks.NETHERRACK);

        noStones.add("if_player");
        noStones.add("else");
        noStones.add("repeat");
        noStones.add("if_var");
        noStones.add("if_entity");
        noStones.add("if_game");

        noChests.add("else");
        noChests.add("call_func");
        noChests.add("event");
        noChests.add("entity_event");

        Minecraft mc = Recode.MC;
        WorldRenderEvents.BEFORE_BLOCK_OUTLINE.register((ctx, outline) -> {
            if (!Config.getBoolean("templatePeeking")) {
                return true;
            }

            BlockState chest = Blocks.CHEST.defaultBlockState();
            chest = chest.setValue(ChestBlock.FACING, Direction.NORTH);

            templatePreview = true;
            templateFits = true;
            try {
                blockRenderer.clear();

                //Save current matrix state
                PoseStack matrix = ctx.matrixStack();
                matrix.pushPose();

                //Offset camera
                Vec3 vec = ctx.camera().getPosition();
                matrix.translate(-vec.x(), -vec.y(), -vec.z());

                //Actual Render

                BlockPos target = new BlockPos(mc.hitResult.getLocation());

                if (!mc.level.getBlockState(target.below()).isAir()
                    && mc.player.isCreative()) {

                    if (TemplateUtils.isTemplate(mc.player.getMainHandItem())) {
                        JsonObject encoded = TemplateUtils.fromItemStack(
                            mc.player.getMainHandItem());
                        JsonArray json = JsonParser.parseString(new String(
                            CompressionUtil.fromGZIP(CompressionUtil.fromBase64(
                                encoded.get("code").getAsString().getBytes()))
                        )).getAsJsonObject().get("blocks").getAsJsonArray();

                        BlockPos dloc = target;

                        for (JsonElement b : json) {
                            JsonObject blocks = b.getAsJsonObject();
                            if (Objects.equals(blocks.get("id").getAsString(), "block")) {
                                String type = blocks.get("block").getAsString();

                                canFit(dloc);
                                blockRenderer.addBlocks(new BlockRenderer.RenderingBlock(dloc, blockTypes.get(type).defaultBlockState(), BlockRenderer.BlockType.BLOCK, true));

                                if (!noChests.contains(type)) {
                                    canFit(dloc.above());
                                    blockRenderer.addBlocks(new BlockRenderer.RenderingBlock(dloc.above(), chest, BlockRenderer.BlockType.CHEST, true));
                                }
                                if (!type.equals("else")) {
                                    ArrayList<CodeBlock> blocknames = ActionDump.getCodeBlock(type);
                                    if(blocknames.size() == 0) return true;
                                    String name = blocknames.get(0).getName();

                                    String action = "";
                                    String subAction = "";
                                    String inverted = "";

                                    if (blocks.has("action")) {
                                        action = blocks.get("action").getAsString();
                                    } else if (blocks.has("data")) {
                                        action = blocks.get("data").getAsString();
                                    }
                                    if (blocks.has("subAction")) {
                                        subAction = blocks.get("subAction").getAsString();
                                    }
                                    if (blocks.has("inverted")) {
                                        inverted = blocks.get("inverted").getAsString();
                                    }
                                    canFit(dloc.west());
                                    blockRenderer.addBlocks(new BlockRenderer.RenderingBlock(dloc.west(), getSign(name, action, subAction, inverted), BlockRenderer.BlockType.SIGN, true));
                                }
                                if (!noStones.contains(type)) {
                                    dloc = dloc.south();
                                    canFit(dloc);
                                    blockRenderer.addBlocks(new BlockRenderer.RenderingBlock(dloc, Blocks.STONE.defaultBlockState(), BlockRenderer.BlockType.BLOCK, true));
                                }

                            } else {
                                boolean open =
                                    Objects.equals(blocks.get("direct").getAsString(), "open");
                                boolean norm =
                                    Objects.equals(blocks.get("type").getAsString(), "norm");

                                BlockState bstate = Blocks.PISTON.defaultBlockState();
                                if (!norm) {
                                    bstate = Blocks.STICKY_PISTON.defaultBlockState();
                                }

                                if (open) {
                                    bstate = bstate.setValue(BlockStateProperties.FACING, Direction.SOUTH);
                                } else {
                                    dloc = dloc.south();
                                }
                                canFit(dloc);
                                blockRenderer.addBlocks(new BlockRenderer.RenderingBlock(dloc, bstate, BlockRenderer.BlockType.BLOCK, true));
                            }
                            dloc = dloc.south();
                        }
                        blockRenderer.render(ctx.world(), matrix, ctx.consumers(), templateFits ? new Color(125, 214, 246, 120) : new Color(250, 135, 135, 140));
                    }
                }

                //Undo changes to matrix
                matrix.popPose();
            } catch (Exception err) {
                err.printStackTrace();
            }
            templatePreview = false;

            return true;
        });
    }

    private SignBlockEntity getSign(String name, String action, String subAction, String inverted) {
        BlockState state = Blocks.OAK_WALL_SIGN.defaultBlockState();
        state = state.setValue(WallSignBlock.FACING, Direction.WEST);

        SignBlockEntity sign = new SignBlockEntity();

        String statename = FabricLoader.getInstance().isDevelopmentEnvironment() ?  "cachedState" : "field_11866";

        try {
            Class<BlockEntity> cl = BlockEntity.class;
            Field f = cl.getDeclaredField(statename);
            f.setAccessible(true);
            f.set(sign,state);
        } catch (Exception err) {
            err.printStackTrace();
        }

        sign.setMessage(0, new TextComponent(name));
        sign.setMessage(1, new TextComponent(action));
        sign.setMessage(2, new TextComponent(subAction));
        sign.setMessage(3, new TextComponent(inverted));

        return sign;
    }

    private void canFit(BlockPos pos) {
        templateFits = Minecraft.getInstance().level.getBlockState(pos).isAir() && templateFits;
    }
}
*/

File: src/main/java/io/github/homchom/recode/feature/rendering/TemplatePeeker.kt
package io.github.homchom.recode.feature.rendering

import com.google.gson.JsonParser
import io.github.homchom.recode.feature.feature
import io.github.homchom.recode.mc
import io.github.homchom.recode.mod.config.Config
import io.github.homchom.recode.sys.hypercube.templates.CompressionUtil
import io.github.homchom.recode.sys.hypercube.templates.TemplateUtil
import net.fabricmc.fabric.api.client.rendering.v1.WorldRenderContext
import net.fabricmc.fabric.api.client.rendering.v1.WorldRenderEvents
import net.minecraft.client.renderer.Sheets
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.network.chat.Component
import net.minecraft.world.level.block.Blocks
import net.minecraft.world.level.block.ChestBlock
import net.minecraft.world.level.block.WallSignBlock
import net.minecraft.world.level.block.entity.ChestBlockEntity
import net.minecraft.world.level.block.entity.SignBlockEntity
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.block.state.properties.BlockStateProperties
import java.awt.Color

val blockTypes = mapOf(
    "func" to Blocks.LAPIS_BLOCK.defaultBlockState(),
    "set_var" to Blocks.IRON_BLOCK.defaultBlockState(),
    "event" to Blocks.DIAMOND_BLOCK.defaultBlockState(),
    "player_action" to Blocks.COBBLESTONE.defaultBlockState(),
    "if_player" to Blocks.OAK_PLANKS.defaultBlockState(),
    "else" to Blocks.END_STONE.defaultBlockState(),
    "call_func" to Blocks.LAPIS_ORE.defaultBlockState(),
    "entity_event" to Blocks.GOLD_BLOCK.defaultBlockState(),
    "process" to Blocks.EMERALD_BLOCK.defaultBlockState(),
    "repeat" to Blocks.PRISMARINE.defaultBlockState(),
    "entity_action" to Blocks.MOSSY_COBBLESTONE.defaultBlockState(),
    "start_process" to Blocks.EMERALD_ORE.defaultBlockState(),
    "if_var" to Blocks.OBSIDIAN.defaultBlockState(),
    "if_entity" to Blocks.BRICKS.defaultBlockState(),
    "control" to Blocks.COAL_BLOCK.defaultBlockState(),
    "select_obj" to Blocks.PURPUR_BLOCK.defaultBlockState(),
    "if_game" to Blocks.RED_NETHER_BRICKS.defaultBlockState(),
    "game_action" to Blocks.NETHERRACK.defaultBlockState(),
)

val blockNames = mapOf(
    "func" to "Function",
    "set_var" to "Set Variable",
    "event" to "Event",
    "player_action" to "Player Action",
    "if_player" to "If Player",
    "else" to "Else",
    "call_func" to "Call Function",
    "entity_event" to "Entity Event",
    "process" to "Process",
    "repeat" to "Repeat",
    "entity_action" to "Entity Action",
    "start_process" to "Start Process",
    "if_var" to "If Variable",
    "if_entity" to "If Entity",
    "control" to "Control",
    "select_obj" to "Select Object",
    "if_game" to "If Game",
    "game_action" to "Game Action"
)

val noStones = setOf("if_player", "else", "repeat", "if_var", "if_entity", "if_game")

val noChests = setOf("else", "call_func", "event", "entity_event")

interface RenderBlock {
    fun render(ctx: WorldRenderContext, color: Color)
    val pos: BlockPos
}

data class NormalRenderBlock(override val pos: BlockPos, val state: BlockState) : RenderBlock {
    override fun render(ctx: WorldRenderContext, color: Color) {
        ctx.matrixStack().pushPose()
        ctx.matrixStack().translate(pos.x.toDouble(), pos.y.toDouble(), pos.z.toDouble())
        mc.blockRenderer.modelRenderer.renderModel(
            ctx.matrixStack().last(),
            ctx.consumers()!!.getBuffer(Sheets.translucentCullBlockSheet()),
            state,
            mc.blockRenderer.getBlockModel(state),
            1F, 1F, 1F,
            16777215,
            655360
        )
        ctx.matrixStack().popPose()
    }
}

data class ChestRenderBlock(override val pos: BlockPos, val state: BlockState) : RenderBlock {
    override fun render(ctx: WorldRenderContext, color: Color) {
        ctx.matrixStack().pushPose()
        ctx.matrixStack().translate(pos.x.toDouble(), pos.y.toDouble(), pos.z.toDouble())

        val chestBlockEntity = ChestBlockEntity(BlockPos.ZERO, state)
        val chestRenderer = mc.blockEntityRenderDispatcher.getRenderer(chestBlockEntity)

        chestRenderer!!.render(
            chestBlockEntity,
            ctx.tickDelta(),
            ctx.matrixStack(),
            mc.renderBuffers().bufferSource(),
            16777215,
            655360
        )

        ctx.matrixStack().popPose()
    }
}

data class SignRenderBlock(override val pos: BlockPos, var state: BlockState, val rows: SignRows) : RenderBlock {
    override fun render(ctx: WorldRenderContext, color: Color) {
        ctx.matrixStack().pushPose()
        ctx.matrixStack().translate(pos.x.toDouble(), pos.y.toDouble(), pos.z.toDouble())

        val signBlockEntity = SignBlockEntity(BlockPos.ZERO, state)

        signBlockEntity.setMessage(0, Component.literal(rows.row0))
        signBlockEntity.setMessage(1, Component.literal(rows.row1))
        signBlockEntity.setMessage(2, Component.literal(rows.row2))
        signBlockEntity.setMessage(3, Component.literal(rows.row3))

        val signRenderer = mc.blockEntityRenderDispatcher.getRenderer(signBlockEntity)

        signRenderer!!.render(
            signBlockEntity,
            ctx.tickDelta(),
            ctx.matrixStack(),
            mc.renderBuffers().bufferSource(),
            16777215,
            655360
        )

        ctx.matrixStack().popPose()
    }
}

data class SignRows(val row0: String, val row1: String, val row2: String, val row3: String)

val FTemplatePeeker = feature("Template Peeker") {
    onLoad {
        WorldRenderEvents.BEFORE_BLOCK_OUTLINE.register { ctx, _ ->
            if (!Config.getBoolean("templatePeeking")
                || mc.player == null
                || !mc.player!!.isCreative
            ) {
                return@register true
            }

            ctx.matrixStack().pushPose()

            val camPos = ctx.camera().position
            ctx.matrixStack().translate(-camPos.x, -camPos.y, -camPos.z)

            try {
                val chestState = Blocks.CHEST.defaultBlockState().setValue(ChestBlock.FACING, Direction.NORTH)
                val todo = mutableSetOf<RenderBlock>()

                val signState = Blocks.OAK_WALL_SIGN.defaultBlockState().setValue(WallSignBlock.FACING, Direction.WEST)

                if (TemplateUtil.isTemplate(mc.player!!.mainHandItem)) {
                    val json = JsonParser.parseString(
                        String(
                            CompressionUtil.fromGZIP(
                                CompressionUtil.fromBase64(
                                    TemplateUtil.read(
                                        mc.player!!.mainHandItem
                                    ).get("code").asString.encodeToByteArray()
                                )
                            )
                        )
                    ).asJsonObject.get("blocks").asJsonArray

                    var currentLoc = BlockPos(mc.hitResult!!.location)

                    if (!mc.level!!.getBlockState(currentLoc.below()).isAir) {
                        for (b in json) {
                            val block = b.asJsonObject
                            if (block["id"].asString == "block") {
                                val type = block["block"].asString

                                todo.add(NormalRenderBlock(currentLoc, blockTypes[type]!!))

                                if (!noChests.contains(type)) {
                                    todo.add(ChestRenderBlock(currentLoc.above(), chestState))
                                }

                                if (type != "else") {
                                    val signRows = SignRows(
                                        blockNames.getOrDefault(type, type),
                                        if (block.has("action")) block["action"].asString else block["data"].asString,
                                        if (block.has("subAction")) block["subAction"].asString else "",
                                        if (block.has("inverted")) block["inverted"].asString else ""
                                    )

                                    todo.add(SignRenderBlock(currentLoc.west(), signState, signRows))
                                }

                                if (!noStones.contains(type)) {
                                    todo.add(NormalRenderBlock(currentLoc.south(), Blocks.STONE.defaultBlockState()))
                                    currentLoc = currentLoc.south()
                                }
                            } else {
                                val open = block["direct"].asString == "open"
                                val norm = block["type"].asString == "norm"

                                var state =
                                    if (norm) Blocks.PISTON.defaultBlockState() else Blocks.STICKY_PISTON.defaultBlockState()

                                if (open) {
                                    state = state.setValue(BlockStateProperties.FACING, Direction.SOUTH)
                                } else {
                                    currentLoc = currentLoc.south()
                                }

                                todo.add(NormalRenderBlock(currentLoc, state))
                            }
                            currentLoc = currentLoc.south()
                        }

                        var fits = true

                        for (renderBlock in todo) {
                            if (!mc.level!!.getBlockState(renderBlock.pos).isAir) {
                                fits = false
                                break
                            }
                        }

                        for (renderBlock in todo) {
                            if (mc.level!!.getBlockState(renderBlock.pos).isAir) {
                                renderBlock.render(
                                    ctx,
                                    if (fits) Color.BLUE else Color.RED
                                )
                            }
                        }
                    }
                }

            } catch (e: Exception) {
                e.printStackTrace()
            }

            ctx.matrixStack().popPose()
            true
        }
    }
}